internal/events/flagged_driver.go

```diff
@@ -0,0 +1,162 @@
+//
+// Copyright 2024 Stacklok, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package events
+
+import (
+	"context"
+	"fmt"
+	"runtime/debug"
+
+	"github.com/ThreeDotsLabs/watermill/message"
+	"github.com/open-feature/go-sdk/openfeature"
+	"github.com/rs/zerolog"
+	"go.opentelemetry.io/otel"
+	"go.opentelemetry.io/otel/attribute"
+	"go.opentelemetry.io/otel/metric"
+
+	"github.com/mindersec/minder/internal/engine/engcontext"
+	"github.com/mindersec/minder/internal/events/common"
+	"github.com/mindersec/minder/internal/flags"
+	serverconfig "github.com/mindersec/minder/pkg/config/server"
+)
+
+type flaggedDriver struct {
+	basePub       message.Publisher
+	experimentPub message.Publisher
+	baseSub       message.Subscriber
+	experimentSub message.Subscriber
+
+	flags             openfeature.IClient
+	publishedMessages metric.Int64Counter
+	readMessages      metric.Int64Counter
+}
+
+// Publish implements message.Publisher.  If the message is flagged by the
+// alternate_message_driver experiment, it is published to the experiment
+// driver, otherwise it is published to the base driver.
+func (f *flaggedDriver) Publish(topic string, messages ...*message.Message) error {
+	// Each message has its own context, so they _could_ be in different flag treatments
+	for _, m := range messages {
+		emptyContext := engcontext.EntityContext{}
+		if engcontext.EntityFromContext(m.Context()) == emptyContext {
+			zerolog.Ctx(m.Context()).Warn().Str("stack", string(debug.Stack())).Msg("No entity in context")
+		}
+		if flags.Bool(m.Context(), f.flags, flags.AlternateMessageDriver) {
+			if err := f.experimentPub.Publish(topic, m); err != nil {
+				return err
+			}
+			f.publishedMessages.Add(m.Context(), 1, metric.WithAttributes(attribute.Bool("experiment", true)))
+		} else {
+			if err := f.basePub.Publish(topic, m); err != nil {
+				return err
+			}
+			f.publishedMessages.Add(m.Context(), 1, metric.WithAttributes(attribute.Bool("experiment", false)))
+		}
+	}
+	return nil
+}
+
+// Subscribe implements message.Subscriber.  In this case, it should subscribe
+// to both the base and experiment drivers, because we might get messages for
+// either.
+func (f *flaggedDriver) Subscribe(ctx context.Context, topic string) (<-chan *message.Message, error) {
+	out := make(chan *message.Message)
```

Comment:
This channel, being unbuffered, will block the pipeline until the already published messages have been consumed (slow subscriber, etc.), which in turn blocks whatever we receive from the other channels (fan-in), basically blocking everything. I'm assuming this is what we have right now (with other implementations, too).

Reply:
Yes, that's on purpose.  (This is called back-pressure.)  We don't want one process (Pod) to greedily read 100 messages from the topic and then spend a long time digesting them while the other processes are sitting around with no work.  By blocking the read-from-queue thing after the first element, it gives a chance for other processes to pick up the next work items.

If we had 4 threads in a process reading messages from the subscriber, we'd end up with each of them pulling one item, and then hitting back-pressure until a thread became free.  I'm not sure whether we have these kinds of worker pools currently in place, but that feels like the right place to manage parallelism to me.

Reply:
I think this should be fine as publishers (what we use today) are not blocked if there are no active subscribers (as it should be, of course, to a limit). 

This is more of a question than a concern: how about something where subscribers can pick multiple events from a queue (in a load-balanced way) and have the ability to process multiple events parallelly (although I think for Minder, we want to sequentially process events that we get from a repo to make sure we are not creating any implicit race conditions (reading different repo config)).

---

I'm fine with this.